File: src/utils/srs.js
// Lightweight spaced-repetition scheduler (SM-2 flavored).
//
// Practice grades a card with one of three buttons: "Got it" (good), "Hard", or
// "Still learning" (again). Each reviewed card stores
//   { ease, interval, reps, lapses, due, last, created, relearn }
// keyed by its Italian term in localStorage. This module is pure and
// fully unit-tested; the React glue lives in src/hooks/useSrs.js.
//
// Two refinements over the original binary version:
//   • a 'hard' grade — a card recalled slowly and one recalled instantly used to
//     advance identically, throwing away information the learner reliably has.
//   • relearning steps — a lapse used to reset straight to the 1d/3d ladder. Now
//     a lapsed card clears a short step (10 min, then 1 day) first, which is
//     standard practice and cuts repeat lapses.

export const DAY = 86400000; // ms in a day
export const MINUTE = 60000;
export const DEFAULT_EASE = 2.5;
export const MIN_EASE = 1.3;
export const DAILY_NEW_CAP = 15; // max brand-new cards introduced per calendar day
export const HARD_FACTOR = 1.2;  // interval multiplier for a 'hard' answer
// Relearning ladder a lapsed card climbs before it is scheduled normally again.
export const RELEARN_STEPS = [10 * MINUTE, DAY];

export const GRADES = ['again', 'hard', 'good'];

// Apply one review to a card's prior state and return its next state.
// `grade` is 'again' | 'hard' | 'good'. `now` is injectable for tests.
export function review(card, grade, now = Date.now()) {
  let ease = card?.ease ?? DEFAULT_EASE;
  let reps = card?.reps ?? 0;
  let lapses = card?.lapses ?? 0;
  let interval = card?.interval ?? 0;
  // Index into RELEARN_STEPS; null/undefined means "not currently relearning".
  let relearn = card?.relearn ?? null;
  // Stamp when a card was first introduced, so the daily new-card cap can count
  // today's new cards across sessions.
  const created = card?.created ?? now;

  if (grade === 'again') {
    // Lower the ease, reset the streak, and put the card at the bottom of the
    // relearning ladder so it comes back within the session.
    reps = 0;
    lapses += 1;
    ease = Math.max(MIN_EASE, ease - 0.2);
    interval = 0;
    relearn = 0;
    return { ease, interval, reps, lapses, due: now + RELEARN_STEPS[0], last: now, created, relearn };
  }

  // A card in relearning climbs one step per correct answer before rejoining
  // the normal ladder.
  if (relearn !== null && relearn !== undefined) {
    const next = relearn + 1;
    if (next < RELEARN_STEPS.length) {
      return { ease, interval, reps, lapses, due: now + RELEARN_STEPS[next], last: now, created, relearn: next };
    }
    relearn = null;
    reps = 1;
    interval = 1;
    return { ease, interval, reps, lapses, due: now + DAY, last: now, created, relearn };
  }

  if (grade === 'hard') {
    // Recalled, but with effort: shrink the ease a little and grow the interval
    // gently rather than by the full ease factor.
    ease = Math.max(MIN_EASE, ease - 0.15);
    reps += 1;
    interval = interval > 0 ? Math.max(1, Math.round(interval * HARD_FACTOR)) : 1;
    return { ease, interval, reps, lapses, due: now + interval * DAY, last: now, created, relearn: null };
  }

  // grade === 'good': advance the interval (1d, 3d, then interval * ease).
  reps += 1;
  if (reps === 1) interval = 1;
  else if (reps === 2) interval = 3;
  else interval = Math.max(1, Math.round(interval * ease));

  return { ease, interval, reps, lapses, due: now + interval * DAY, last: now, created, relearn: null };
}

export function isDue(card, now = Date.now()) {
  return !!card && card.due <= now;
}

// Build a practice queue from the full (already phase-filtered) card list and
// the SRS store. Due cards come first (earliest due first), then up to `newCap`
// never-seen cards, capped at `maxSession` total.
export function buildQueue(cards, store, { now = Date.now(), newCap = 12, maxSession = 20 } = {}) {
  const due = [];
  const fresh = [];
  for (const c of cards) {
    const st = store[c.it];
    if (!st) fresh.push(c);
    else if (st.due <= now) due.push({ c, due: st.due });
  }
  due.sort((a, b) => a.due - b.due);
  const dueCards = due.map((d) => d.c);
  const newCards = fresh.slice(0, newCap);
  return [...dueCards, ...newCards].slice(0, maxSession);
}

function sameLocalDay(a, b) {
  const da = new Date(a);
  const db = new Date(b);
  return da.getFullYear() === db.getFullYear() && da.getMonth() === db.getMonth() && da.getDate() === db.getDate();
}

// How many brand-new cards were first introduced on the same calendar day as
// `now` — drives the daily new-card cap.
export function newIntroducedToday(store, now = Date.now()) {
  let count = 0;
  for (const key in store) {
    const c = store[key];
    if (c && c.created && sameLocalDay(c.created, now)) count += 1;
  }
  return count;
}

// The new-card allowance left for today given the daily cap.
export function newAllowanceToday(store, now = Date.now(), cap = DAILY_NEW_CAP) {
  return Math.max(0, cap - newIntroducedToday(store, now));
}

// Counts for the start screen: how many are due now, never seen, or already
// learned (seen at least once).
export function stats(cards, store, now = Date.now()) {
  let due = 0;
  let fresh = 0;
  let learned = 0;
  for (const c of cards) {
    const st = store[c.it];
    if (!st) {
      fresh += 1;
    } else {
      learned += 1;
      if (st.due <= now) due += 1;
    }
  }
  return { due, new: fresh, learned, total: cards.length };
}

File: ios-native/BibbiaCore/Sources/BibbiaCore/SRS.swift
import Foundation

// Port of src/utils/srs.js — the SM-2-flavored spaced-repetition scheduler. Timestamps are **milliseconds since the epoch** (JS Date.now()),
// and the card JSON shape matches the web app's `italian-bible-srs` store
// exactly, so sync-snapshot backups move between the PWA and this app.

public let MS_PER_DAY: Double = 86_400_000
public let DEFAULT_EASE = 2.5
public let MIN_EASE = 1.3
public let DAILY_NEW_CAP = 15  // max brand-new cards introduced per calendar day
public let MS_PER_MINUTE: Double = 60_000
public let HARD_FACTOR = 1.2   // interval multiplier for a 'hard' answer
/// Relearning ladder a lapsed card climbs before it is scheduled normally again.
public let RELEARN_STEPS: [Double] = [10 * MS_PER_MINUTE, MS_PER_DAY]

public struct SRSCard: Codable, Equatable {
    public var ease: Double
    public var interval: Double   // days
    public var reps: Int
    public var lapses: Int
    public var due: Double        // ms epoch
    public var last: Double?      // ms epoch
    public var created: Double?   // ms epoch (older web cards may lack it)
    /// Index into RELEARN_STEPS; nil means "not currently relearning".
    public var relearn: Int?

    public init(ease: Double = DEFAULT_EASE, interval: Double = 0, reps: Int = 0,
                lapses: Int = 0, due: Double = 0, last: Double? = nil, created: Double? = nil,
                relearn: Int? = nil) {
        self.ease = ease
        self.interval = interval
        self.reps = reps
        self.lapses = lapses
        self.due = due
        self.last = last
        self.created = created
        self.relearn = relearn
    }
}

/// The web app's whole SRS store: Italian term → card state.
public typealias SRSStore = [String: SRSCard]

public enum SRSGrade: String {
    case good   // "Got it"
    case hard   // "Hard" — recalled, but with effort
    case again  // "Still learning"
}

/// Apply one review to a card's prior state and return its next state.
public func srsReview(_ card: SRSCard?, grade: SRSGrade, now: Double) -> SRSCard {
    var ease = card?.ease ?? DEFAULT_EASE
    var reps = card?.reps ?? 0
    var lapses = card?.lapses ?? 0
    var interval = card?.interval ?? 0
    let relearn = card?.relearn
    // Stamp when a card was first introduced, so the daily new-card cap can
    // count today's new cards across sessions.
    let created = card?.created ?? now

    if grade == .again {
        // Lower the ease, reset the streak, and put the card at the bottom of
        // the relearning ladder so it comes back within the session.
        reps = 0
        lapses += 1
        ease = max(MIN_EASE, ease - 0.2)
        interval = 0
        return SRSCard(ease: ease, interval: interval, reps: reps, lapses: lapses,
                       due: now + RELEARN_STEPS[0], last: now, created: created, relearn: 0)
    }

    // A card in relearning climbs one step per correct answer before rejoining
    // the normal ladder.
    if let step = relearn {
        let next = step + 1
        if next < RELEARN_STEPS.count {
            return SRSCard(ease: ease, interval: interval, reps: reps, lapses: lapses,
                           due: now + RELEARN_STEPS[next], last: now, created: created, relearn: next)
        }
        return SRSCard(ease: ease, interval: 1, reps: 1, lapses: lapses,
                       due: now + MS_PER_DAY, last: now, created: created, relearn: nil)
    }

    if grade == .hard {
        // Recalled, but with effort: shrink the ease a little and grow the
        // interval gently rather than by the full ease factor.
        ease = max(MIN_EASE, ease - 0.15)
        reps += 1
        interval = interval > 0 ? max(1, (interval * HARD_FACTOR).rounded()) : 1
        return SRSCard(ease: ease, interval: interval, reps: reps, lapses: lapses,
                       due: now + interval * MS_PER_DAY, last: now, created: created, relearn: nil)
    }

    // grade == .good: advance the interval (1d, 3d, then interval * ease).
    reps += 1
    if reps == 1 { interval = 1 }
    else if reps == 2 { interval = 3 }
    else { interval = max(1, (interval * ease).rounded()) }

    return SRSCard(ease: ease, interval: interval, reps: reps, lapses: lapses,
                   due: now + interval * MS_PER_DAY, last: now, created: created, relearn: nil)
}

public func isDue(_ card: SRSCard?, now: Double) -> Bool {
    guard let card else { return false }
    return card.due <= now
}

/// Build a practice queue from the (already-filtered) card list and the SRS
/// store. Due cards first (earliest due first), then up to `newCap` never-seen
/// cards, capped at `maxSession` total.
public func buildQueue(cards: [VocabCard], store: SRSStore, now: Double,
                       newCap: Int = 12, maxSession: Int = 20) -> [VocabCard] {
    var due: [(card: VocabCard, due: Double, order: Int)] = []
    var fresh: [VocabCard] = []
    for (i, c) in cards.enumerated() {
        if let st = store[c.it] {
            if st.due <= now { due.append((c, st.due, i)) }
        } else {
            fresh.append(c)
        }
    }
    // Earliest due first; original order breaks ties (JS Array.sort is stable).
    due.sort { $0.due != $1.due ? $0.due < $1.due : $0.order < $1.order }
    let queue = due.map(\.card) + Array(fresh.prefix(newCap))
    return Array(queue.prefix(maxSession))
}

func sameLocalDay(_ a: Double, _ b: Double, calendar: Calendar) -> Bool {
    calendar.isDate(Date(timeIntervalSince1970: a / 1000),
                    inSameDayAs: Date(timeIntervalSince1970: b / 1000))
}

/// How many brand-new cards were first introduced on the same calendar day as
/// `now` — drives the daily new-card cap.
public func newIntroducedToday(store: SRSStore, now: Double,
                               calendar: Calendar = .current) -> Int {
    store.values.filter { card in
        guard let created = card.created else { return false }
        return sameLocalDay(created, now, calendar: calendar)
    }.count
}

/// The new-card allowance left for today given the daily cap.
public func newAllowanceToday(store: SRSStore, now: Double,
                              cap: Int = DAILY_NEW_CAP,
                              calendar: Calendar = .current) -> Int {
    max(0, cap - newIntroducedToday(store: store, now: now, calendar: calendar))
}

public struct SRSStats: Equatable {
    public let due: Int
    public let new: Int
    public let learned: Int
    public let total: Int
}

/// Counts for the practice start screen.
public func srsStats(cards: [VocabCard], store: SRSStore, now: Double) -> SRSStats {
    var due = 0
    var fresh = 0
    var learned = 0
    for c in cards {
        if let st = store[c.it] {
            learned += 1
            if st.due <= now { due += 1 }
        } else {
            fresh += 1
        }
    }
    return SRSStats(due: due, new: fresh, learned: learned, total: cards.count)
}
